Avoid unnecessary lazy init in StructuralComparisons

stephentoub opened the pull request:
The only benefit of having these be lazy in this manner is avoiding one of the two objects being constructed. We can instead just combine them into a singleton that's initialized in a static readonly.

Best reviewed without whitespace diff: https://github.com/dotnet/runtime/pull/101344/files?w=1

jkotas reviewed (approved):
- Some code out there may get broken by `StructuralComparisons.StructuralComparer` implementing `IEqualityComparer` or `StructuralComparisons.StructuralEqualityComparer` implementing `IComparer`.
- IL trimmer may end up keeping unusued interface implementation around now that both interfaces are implemented by the same type.

Structural equality is used very rarely, so it is unlikely for these concerns to be a problem in practice. I am mentioning them out of abundance of caution. An alternative would be to keep the two types and add static readonly field in each of the types to get the lazy initialization for free.

stephentoub commented:
> An alternative would be to keep the two types and add static readonly field in each of the types

I'll switch to that

stephentoub committed:
Avoid unnecessary lazy init in StructuralComparisons

Move the fields to the target types and make them readonly instead of manually lazy.

stephentoub committed:
Update StructuralComparisons.cs
